pkg/filesystem/nydus/fs.go

```diff
@@ -91,12 +127,118 @@ func (fs *filesystem) newSharedDaemon() (*daemon.Daemon, error) {
 }
 
 func (fs *filesystem) Support(ctx context.Context, labels map[string]string) bool {
-	_, ok := labels[label.NydusDataLayer]
-	return ok
+	_, dataOk := labels[label.NydusDataLayer]
+	_, metaOk := labels[label.NydusMetaLayer]
+	return dataOk || metaOk
+}
+
+func isRafsV6(buf []byte) bool {
+	return nativeEndian.Uint32(buf[RafsV6SuppeOffset:]) == RafsV6Magic
+}
+
+func getBootstrapRealSizeInV6(buf []byte) uint64 {
+	return nativeEndian.Uint64(buf[ChunkInfoOffset:])
 }
 
-func (fs *filesystem) PrepareLayer(context.Context, storage.Snapshot, map[string]string) error {
-	panic("implement me")
+func writeBootstrapToFile(reader io.Reader, bootstrap *os.File, LegacyBootstrap *os.File) error {
+	start := time.Now()
+	defer func() {
+		duration := time.Since(start)
+		log.L.Infof("read bootstrap duration %d", duration.Milliseconds())
+	}()
+	rd, err := gzip.NewReader(reader)
+	if err != nil {
+		return err
+	}
+	found := false
+	tr := tar.NewReader(rd)
+	var finalBootstarp *os.File
+	for {
+		h, err := tr.Next()
+		if err != nil {
+			return err
+		}
+		if h.Name == BootstrapFile {
+			found = true
+			finalBootstarp = bootstrap
+			break
+		}
+
+		if h.Name == LegacyBootstrapFile {
+			found = true
+			finalBootstarp = LegacyBootstrap
+			break
+		}
+	}
+
+	if !found {
+		return fmt.Errorf("not found file image.boot in targz")
+	}
+	buf := make([]byte, MaxSuperBlockSize)
+	_, err = tr.Read(buf)
+	if err != nil {
+		return err
+	}
+	_, err = finalBootstarp.Write(buf)
+
+	if err != nil {
+		return err
+	}
+
+	if isRafsV6(buf) {
+		size := getBootstrapRealSizeInV6(buf)
+		if size < MaxSuperBlockSize {
+			return fmt.Errorf("invalid chunk info offset %d", size)
```

Comment:
I get that the start position of chunk info is the size of bootstrap, but can we put "invalid bootstrap size" for the error message instead?

Reply:
fixed